File: examples/showcase.php
<?php
/**
 * Hatsune Miku Theme - PHP Showcase
 * All-Miku Synthesis: Every voice, one stage.
 */

declare(strict_types=1);

namespace MikuTheme\Showcase;

use Exception;
use Generator;
use ArrayAccess;
use JsonSerializable;
use Stringable;

// Constants: #E05096 (Magenta LED - Rhythm)
const CANONICAL_COLOR = '#39C5BB';
const DEFAULT_BPM = 39;
const MAX_ENERGY = 100;

define('MIKU_MODEL_ID', 'CV01');

// Interface: #B2EBE7 Bold (Ice Teal - Append Light)
interface Synthesizer
{
    // Method signature: #00BCD4 (NT Cyan)
    public function sing(string $song): string;
    public function getVersion(): MikuVersion;
}

// Backed enum (PHP 8.1+): #B2EBE7 Bold, EnumMember: #E05096
enum MikuVersion: string
{
    case V2Classic = 'V2_CLASSIC';
    case V3 = 'V3';
    case V4X = 'V4X';
    case NT = 'NT';
    case Sekai = 'SEKAI';
    case V6AI = 'V6_AI';

    // Enum method
    public function getReleaseYear(): int
    {
        return match ($this) {
            self::V2Classic => 2007,
            self::V3 => 2013,
            self::V4X => 2016,
            self::NT, self::Sekai => 2020,
            self::V6AI => 2024,
        };
    }
}

// Unit enum
enum StageMode
{
    case Idle;
    case Performing;
    case Holographic;
}

// Attribute (PHP 8+): #FFD740 Italic (Meta Amber)
#[Attribute(Attribute::TARGET_CLASS | Attribute::TARGET_METHOD)]
class Vocaloid
{
    public function __construct(
        public readonly string $engineName = 'VOCALOID',
        public readonly int $releaseYear = 2007
    ) {}
}

// Custom exception: #FF5370
class PerformanceException extends Exception
{
    public function __construct(
        string $message,
        public readonly ?string $song = null
    ) {
        parent::__construct($message);
    }
}

// Trait
trait EnergyManagement
{
    private int $energy = MAX_ENERGY;

    public function getEnergy(): int
    {
        return $this->energy;
    }

    public function setEnergy(int $value): void
    {
        $this->energy = max(0, min(MAX_ENERGY, $value));
    }

    public function consumeEnergy(int $amount): void
    {
        $this->energy = max(0, $this->energy - $amount);
    }
}

// Readonly class (PHP 8.2+)
readonly class VoiceBank implements JsonSerializable
{
    // Constructor property promotion
    public function __construct(
        public string $name,
        public MikuVersion $version,
        public array $frequencyRange = [80, 1100]
    ) {}

    public function jsonSerialize(): array
    {
        return [
            'name' => $this->name,
            'version' => $this->version->value,
            'frequencyRange' => $this->frequencyRange,
        ];
    }
}

// Class: #B2EBE7 Bold (Ice Teal)
#[Vocaloid(engineName: 'VOCALOID2', releaseYear: 2007)]
class DigitalDiva implements Synthesizer, ArrayAccess, Stringable
{
    use EnergyManagement;

    // Static property: #90B8B2 Underline
    public static string $modelId = 'CV01';

    // Class constants
    private const VOICE_TYPES = ['dark', 'soft', 'light', 'sweet', 'vivid', 'solid'];

    // Properties: #90B8B2
    private StageMode $mode = StageMode::Idle;
    private array $metadata = [];

    // Constructor with property promotion
    public function __construct(
        public readonly string $name = 'Hatsune Miku',
        public readonly VoiceBank $voiceBank = new VoiceBank('Hatsune Miku', MikuVersion::V2Classic)
    ) {}

    // Method: #00BCD4 (NT Cyan)
    public function sing(string $song): string
    {
        // Keywords: #39C5BB Bold
        if ($this->energy < 10) {
            throw new PerformanceException(
                'Low energy: Please recharge with leeks',
                $song
            );
        }

        // String: #9CCC65 (Negi Green)
        // Variable in string: #FF80AB (Soft Pink)
        $message = "[MIKU] Now singing: {$song}";
        echo $message . PHP_EOL;

        $this->consumeEnergy(10);
        return $message;
    }

    public function getVersion(): MikuVersion
    {
        return $this->voiceBank->version;
    }

    // Static method: #00BCD4 Underline
    public static function getCanonicalColor(): string
    {
        return CANONICAL_COLOR;
    }

    // Magic method: #00BCD4 Bold
    public function __toString(): string
    {
        return sprintf('%s (%s)', $this->name, $this->voiceBank->version->value);
    }

    public function __get(string $name): mixed
    {
        return $this->metadata[$name] ?? null;
    }

    public function __set(string $name, mixed $value): void
    {
        $this->metadata[$name] = $value;
    }

    public function __isset(string $name): bool
    {
        return isset($this->metadata[$name]);
    }

    // ArrayAccess implementation
    public function offsetExists(mixed $offset): bool
    {
        return isset(self::VOICE_TYPES[$offset]);
    }

    public function offsetGet(mixed $offset): mixed
    {
        return self::VOICE_TYPES[$offset] ?? null;
    }

    public function offsetSet(mixed $offset, mixed $value): void
    {
        // Readonly - do nothing
    }

    public function offsetUnset(mixed $offset): void
    {
        // Readonly - do nothing
    }

    // Generator method
    public function voiceTypeGenerator(): Generator
    {
        foreach (self::VOICE_TYPES as $index => $type) {
            yield $index => $type;
        }
    }
}

// First-class callable syntax (PHP 8.1+)
$processor = strtoupper(...);

// Named arguments
$miku = new DigitalDiva(
    name: 'Hatsune Miku',
    voiceBank: new VoiceBank(
        name: 'Miku',
        version: MikuVersion::V2Classic,
        frequencyRange: [80, 1100]
    )
);

// Arrow function: #00BCD4
$getUpperSong = fn(string $song): string => strtoupper($song);

// Closure
$singWithPrefix = function (string $song) use ($miku): string {
    return "[{$miku->name}] " . $song;
};

// Array operations
$songs = ['Melt', 'World is Mine', 'Rolling Girl'];

// Array map with arrow function
$upperSongs = array_map(fn($s) => strtoupper($s), $songs);

// Array filter
$longSongs = array_filter($songs, fn($s) => strlen($s) > 4);

// Array reduce
$totalLength = array_reduce($songs, fn($carry, $s) => $carry + strlen($s), 0);

// Match expression (PHP 8+)
$description = match ($miku->getVersion()) {
    MikuVersion::V2Classic => 'The original classic',
    MikuVersion::NT => 'The modern NT engine',
    MikuVersion::Sekai => 'Project SEKAI version',
    default => 'Unknown version',
};

// Null-safe operator (PHP 8+)
$versionValue = $miku->voiceBank?->version?->value;

// Try-catch with specific types
try {
    foreach ($songs as $song) {
        $miku->sing($song);
    }
} catch (PerformanceException $e) {
    // Error: #FF5370
    echo "Performance error: {$e->getMessage()}" . PHP_EOL;
    if ($e->song !== null) {
        echo "Failed on: {$e->song}" . PHP_EOL;
    }
} finally {
    echo "Performance attempt completed" . PHP_EOL;
}

// Heredoc syntax
$json = <<<JSON
{
    "name": "Hatsune Miku",
    "version": "V2",
    "color": "#39C5BB"
}
JSON;

// Nowdoc syntax (no variable interpolation)
$template = <<<'TEMPLATE'
[MIKU] Template without $variable interpolation
TEMPLATE;

// Spread operator
$allSongs = [...$songs, 'Love is War', 'Ievan Polkka'];

// Union types and intersection types (PHP 8+)
function process(string|int $input): string|int
{
    return is_string($input) ? strtoupper($input) : $input * 2;
}

echo "Color: " . DigitalDiva::getCanonicalColor() . PHP_EOL;
echo "Diva: " . $miku . PHP_EOL;
echo $description . PHP_EOL;

File: examples/showcase.go
// Hatsune Miku Theme - Go Showcase
// All-Miku Synthesis: Every voice, one stage.

package main

import (
	"context"
	"errors"
	"fmt"
	"sync"
	"time"
)

// Constants: #E05096 (Magenta LED - Rhythm)
const (
	CanonicalColor = "#39C5BB"
	DefaultBPM     = 39
	MaxEnergy      = 100
)

// Package-level variables
var (
	// Error types: #FF5370
	ErrLowEnergy    = errors.New("low energy: please recharge with leeks")
	ErrStageNotReady = errors.New("stage not ready")
)

// Interface: #B2EBE7 Bold (Ice Teal - Append Light)
type Synthesizer interface {
	// Method signature: #00BCD4 (NT Cyan)
	Sing(ctx context.Context, song string) error
	GetVersion() MikuVersion
}

// Type alias
type MikuVersion string

// Enum-like constants
const (
	V2Classic MikuVersion = "V2_CLASSIC"
	V3        MikuVersion = "V3"
	V4X       MikuVersion = "V4X"
	NT        MikuVersion = "NT"
	Sekai     MikuVersion = "SEKAI"
	V6AI      MikuVersion = "V6_AI"
)

// Struct: #B2EBE7 Bold (Ice Teal)
type VoiceBank struct {
	Name           string         // Field: #90B8B2
	Version        MikuVersion
	FrequencyRange [2]int
}

// Main struct with generics (Go 1.18+)
// Generic type T: #B2EBE7 Italic
type DigitalDiva[T any] struct {
	name      string
	energy    int
	voiceBank VoiceBank
	metadata  map[string]T
	mu        sync.RWMutex
}

// Constructor function: #00BCD4
func NewDigitalDiva[T any](name string, vb VoiceBank) *DigitalDiva[T] {
	return &DigitalDiva[T]{
		name:      name,
		energy:    MaxEnergy,
		voiceBank: vb,
		metadata:  make(map[string]T),
	}
}

// Method with receiver: #E05096 Italic (receiver variable)
// Method name: #00BCD4 (NT Cyan)
func (d *DigitalDiva[T]) Sing(ctx context.Context, song string) error {
	d.mu.RLock()
	energy := d.energy
	d.mu.RUnlock()

	// Keywords: #39C5BB Bold
	if energy < 10 {
		return ErrLowEnergy
	}

	// Select statement: #39C5BB Bold
	select {
	case <-ctx.Done():
		return ctx.Err()
	default:
		// String formatting: #9CCC65 (Negi Green)
		fmt.Printf("[MIKU] Now singing: %s\n", song)
	}

	// Switch statement: #39C5BB Bold
	switch d.voiceBank.Version {
	case V2Classic:
		return d.classicVoice(song)
	case NT:
		return d.ntVoice(song)
	case Sekai:
		return d.sekaiVoice(song)
	default:
		return d.defaultVoice(song)
	}
}

func (d *DigitalDiva[T]) GetVersion() MikuVersion {
	return d.voiceBank.Version
}

func (d *DigitalDiva[T]) classicVoice(song string) error {
	fmt.Printf("V2 Classic: %s\n", song)
	return nil
}

func (d *DigitalDiva[T]) ntVoice(song string) error {
	fmt.Printf("NT Modern: %s\n", song)
	return nil
}

func (d *DigitalDiva[T]) sekaiVoice(song string) error {
	fmt.Printf("SEKAI Stage: %s\n", song)
	return nil
}

func (d *DigitalDiva[T]) defaultVoice(song string) error {
	fmt.Printf("Default: %s\n", song)
	return nil
}

// Method to update energy with mutex
func (d *DigitalDiva[T]) ConsumeEnergy(amount int) {
	d.mu.Lock()
	defer d.mu.Unlock() // Keyword: #39C5BB Bold

	d.energy -= amount
	if d.energy < 0 {
		d.energy = 0
	}
}

// Goroutine example
func (d *DigitalDiva[T]) PerformAsync(ctx context.Context, songs []string) <-chan error {
	// Channel: return type
	errCh := make(chan error, len(songs))

	// Goroutine: #39C5BB Bold
	go func() {
		defer close(errCh)

		for _, song := range songs {
			// Range: #39C5BB Bold
			if err := d.Sing(ctx, song); err != nil {
				errCh <- err
				return
			}
			time.Sleep(100 * time.Millisecond)
		}
	}()

	return errCh
}

// AppendVoice types
type AppendType int

const (
	Dark AppendType = iota
	Soft
	Light
	Sweet
	Vivid
	Solid
)

// Stringer interface implementation
func (a AppendType) String() string {
	// Array literal
	names := [...]string{"dark", "soft", "light", "sweet", "vivid", "solid"}

	if a < Dark || a > Solid {
		return "unknown"
	}
	return names[a]
}

// Function type
type SongProcessor func(song string) (string, error)

// Higher-order function
func WithLogging(processor SongProcessor) SongProcessor {
	return func(song string) (string, error) {
		fmt.Printf("Processing: %s\n", song)
		result, err := processor(song)
		if err != nil {
			fmt.Printf("Error: %v\n", err)
		}
		return result, err
	}
}

// Main function
func main() {
	// Variable declaration: #39C5BB Bold (var/const)
	voiceBank := VoiceBank{
		Name:           "Hatsune Miku",
		Version:        V2Classic,
		FrequencyRange: [2]int{80, 1100},
	}

	// Type parameter: #B2EBE7
	miku := NewDigitalDiva[string]("Miku", voiceBank)

	// Context: #B2EBE7 (Type)
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	// Error handling
	if err := miku.Sing(ctx, "World is Mine"); err != nil {
		fmt.Printf("Error: %v\n", err)
		return
	}

	// Slice literal
	songs := []string{"Melt", "Love is War", "Rolling Girl"}

	// Channel receive
	for err := range miku.PerformAsync(ctx, songs) {
		if err != nil {
			fmt.Printf("Performance error: %v\n", err)
		}
	}

	fmt.Println("Performance complete!")
}
